File: src/calculators/statistics.py
"""
통계 계산 모듈
평균, 중앙값, 최빈값, 분산, 표준편차 등을 계산합니다.
"""
import math
from typing import List, Dict, Tuple, Optional
from dataclasses import dataclass
from collections import Counter
from ..utils.logger import get_logger

logger = get_logger()


@dataclass
class StatisticsResult:
    """통계 결과 클래스"""
    data: List[float]
    mean: float  # 평균
    median: float  # 중앙값
    mode: Optional[List[float]]  # 최빈값 (여러 개 가능)
    variance: float  # 분산
    std_dev: float  # 표준편차
    range_value: float  # 범위
    quartiles: Tuple[float, float, float]  # Q1, Q2(median), Q3
    steps: List[str]


class StatisticsCalculator:
    """통계 계산기 클래스"""

    def __init__(self):
        """초기화"""
        logger.info("통계 계산기 초기화")

    def calculate_all(self, data: List[float]) -> StatisticsResult:
        """
        모든 통계량 계산

        Args:
            data: 데이터 리스트

        Returns:
            StatisticsResult 객체
        """
        logger.debug(f"통계 계산: {len(data)}개 데이터")

        if not data:
            raise ValueError("데이터가 비어있습니다.")

        steps = []
        steps.append(f"주어진 데이터: {data}")
        steps.append(f"데이터 개수: {len(data)}")

        # 평균
        mean = self.calculate_mean(data)
        steps.append(f"\n평균 = {sum(data)} / {len(data)} = {mean}")

        # 중앙값
        median = self.calculate_median(data)
        sorted_data = sorted(data)
        steps.append(f"\n정렬된 데이터: {sorted_data}")
        steps.append(f"중앙값 = {median}")

        # 최빈값
        mode = self.calculate_mode(data)
        if mode:
            steps.append(f"\n최빈값: {mode}")
        else:
            steps.append("\n최빈값: 없음 (모든 값의 빈도가 같음)")

        # 분산
        variance = self.calculate_variance(data, mean)
        steps.append(f"\n분산 = {variance}")

        # 표준편차
        std_dev = self.calculate_std_dev(variance)
        steps.append(f"표준편차 = √{variance} = {std_dev}")

        # 범위
        range_value = max(data) - min(data)
        steps.append(f"\n범위 = {max(data)} - {min(data)} = {range_value}")

        # 사분위수
        q1, q2, q3 = self.calculate_quartiles(data)
        steps.append(f"\n사분위수:")
        steps.append(f"  Q1 (제1사분위수) = {q1}")
        steps.append(f"  Q2 (제2사분위수, 중앙값) = {q2}")
        steps.append(f"  Q3 (제3사분위수) = {q3}")
        steps.append(f"  IQR (사분위수 범위) = {q3 - q1}")

        logger.info(f"통계 계산 완료: 평균={mean}, 중앙값={median}")

        return StatisticsResult(
            data=data,
            mean=mean,
            median=median,
            mode=mode,
            variance=variance,
            std_dev=std_dev,
            range_value=range_value,
            quartiles=(q1, q2, q3),
            steps=steps
        )

    def calculate_mean(self, data: List[float]) -> float:
        """평균 계산"""
        if not data:
            raise ValueError("데이터가 비어있습니다.")
        return sum(data) / len(data)

    def calculate_median(self, data: List[float]) -> float:
        """중앙값 계산"""
        if not data:
            raise ValueError("데이터가 비어있습니다.")

        sorted_data = sorted(data)
        n = len(sorted_data)

        if n % 2 == 0:
            # 짝수 개: 중간 두 값의 평균
            return (sorted_data[n // 2 - 1] + sorted_data[n // 2]) / 2
        else:
            # 홀수 개: 중간 값
            return sorted_data[n // 2]

    def calculate_mode(self, data: List[float]) -> Optional[List[float]]:
        """
        최빈값 계산

        Returns:
            최빈값 리스트 (여러 개일 수 있음) 또는 None
        """
        if not data:
            raise ValueError("데이터가 비어있습니다.")

        counter = Counter(data)
        max_count = max(counter.values())

        # 모든 값의 빈도가 같으면 최빈값 없음
        if max_count == 1:
            return None

        # 최대 빈도를 가진 값들
        modes = [value for value, count in counter.items() if count == max_count]
        return sorted(modes)

    def calculate_variance(self, data: List[float], mean: Optional[float] = None) -> float:
        """
        분산 계산

        Args:
            data: 데이터 리스트
            mean: 평균 (미리 계산된 경우)

        Returns:
            분산
        """
        if not data:
            raise ValueError("데이터가 비어있습니다.")

        if mean is None:
            mean = self.calculate_mean(data)

        # 분산 = Σ(x - mean)² / n
        squared_diffs = [(x - mean) ** 2 for x in data]
        variance = sum(squared_diffs) / len(data)

        return variance

    def calculate_std_dev(self, variance: float) -> float:
        """표준편차 계산"""
        return math.sqrt(variance)

    def calculate_quartiles(self, data: List[float]) -> Tuple[float, float, float]:
        """
        사분위수 계산

        Returns:
            (Q1, Q2, Q3)
        """
        if not data:
            raise ValueError("데이터가 비어있습니다.")

        sorted_data = sorted(data)
        n = len(sorted_data)

        # Q2 (중앙값)
        q2 = self.calculate_median(sorted_data)

        # Q1 (하위 50%의 중앙값)
        if n % 2 == 0:
            lower_half = sorted_data[:n // 2]
        else:
            lower_half = sorted_data[:n // 2]
        q1 = self.calculate_median(lower_half) if lower_half else sorted_data[0]

        # Q3 (상위 50%의 중앙값)
        if n % 2 == 0:
            upper_half = sorted_data[n // 2:]
        else:
            upper_half = sorted_data[n // 2 + 1:]
        q3 = self.calculate_median(upper_half) if upper_half else sorted_data[-1]

        return (q1, q2, q3)

    def calculate_frequency_table(self, data: List[float]) -> Dict[float, int]:
        """도수분포표 생성"""
        if not data:
            raise ValueError("데이터가 비어있습니다.")

        return dict(Counter(data))

    def calculate_z_score(self, value: float, mean: float, std_dev: float) -> float:
        """
        z-점수 (표준점수) 계산

        Args:
            value: 데이터 값
            mean: 평균
            std_dev: 표준편차

        Returns:
            z-점수
        """
        if std_dev == 0:
            raise ValueError("표준편차가 0입니다.")

        return (value - mean) / std_dev

File: crates/math-core/src/statistics.rs
//! Statistical calculations

use crate::{errors::*, validation::*};
use std::collections::HashMap;

/// Calculate arithmetic mean
pub fn calculate_mean(data: &[f64]) -> Result<f64> {
    if data.is_empty() {
        return Err(MathError::InvalidInput("Data cannot be empty".to_string()));
    }

    for (i, &val) in data.iter().enumerate() {
        validate_number(val, &format!("data[{}]", i))?;
    }

    let sum: f64 = data.iter().sum();
    Ok(sum / data.len() as f64)
}

/// Calculate median
pub fn calculate_median(data: &[f64]) -> Result<f64> {
    if data.is_empty() {
        return Err(MathError::InvalidInput("Data cannot be empty".to_string()));
    }

    for (i, &val) in data.iter().enumerate() {
        validate_number(val, &format!("data[{}]", i))?;
    }

    let mut sorted = data.to_vec();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());

    let len = sorted.len();
    if len % 2 == 0 {
        Ok((sorted[len / 2 - 1] + sorted[len / 2]) / 2.0)
    } else {
        Ok(sorted[len / 2])
    }
}

/// Calculate mode (returns all modes if multiple exist)
pub fn calculate_mode(data: &[f64]) -> Result<Vec<f64>> {
    if data.is_empty() {
        return Err(MathError::InvalidInput("Data cannot be empty".to_string()));
    }

    for (i, &val) in data.iter().enumerate() {
        validate_number(val, &format!("data[{}]", i))?;
    }

    let mut frequency: HashMap<i64, usize> = HashMap::new();

    // Convert to i64 for HashMap key (assuming reasonable precision)
    for &val in data {
        let key = (val * 1000.0).round() as i64;
        *frequency.entry(key).or_insert(0) += 1;
    }

    let max_count = frequency.values().max().unwrap();
    let modes: Vec<f64> = frequency
        .iter()
        .filter(|(_, &count)| count == *max_count)
        .map(|(key, _)| *key as f64 / 1000.0)
        .collect();

    Ok(modes)
}

/// Calculate variance (population or sample)
pub fn calculate_variance(data: &[f64], sample: bool) -> Result<f64> {
    if data.is_empty() {
        return Err(MathError::InvalidInput("Data cannot be empty".to_string()));
    }

    if sample && data.len() < 2 {
        return Err(MathError::InvalidInput(
            "Sample variance requires at least 2 data points".to_string()
        ));
    }

    let mean = calculate_mean(data)?;
    let squared_diffs: f64 = data.iter()
        .map(|&x| (x - mean).powi(2))
        .sum();

    let divisor = if sample {
        data.len() - 1
    } else {
        data.len()
    };

    Ok(squared_diffs / divisor as f64)
}

/// Calculate standard deviation
pub fn calculate_std_dev(data: &[f64], sample: bool) -> Result<f64> {
    let variance = calculate_variance(data, sample)?;
    Ok(variance.sqrt())
}

/// Calculate range (max - min)
pub fn calculate_range(data: &[f64]) -> Result<f64> {
    if data.is_empty() {
        return Err(MathError::InvalidInput("Data cannot be empty".to_string()));
    }

    for (i, &val) in data.iter().enumerate() {
        validate_number(val, &format!("data[{}]", i))?;
    }

    let min = data.iter().fold(f64::INFINITY, |a, &b| a.min(b));
    let max = data.iter().fold(f64::NEG_INFINITY, |a, &b| a.max(b));

    Ok(max - min)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::EPSILON;

    #[test]
    fn test_mean() {
        let data = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        let mean = calculate_mean(&data).unwrap();
        assert!((mean - 3.0).abs() < EPSILON);
    }

    #[test]
    fn test_median_odd() {
        let data = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        let median = calculate_median(&data).unwrap();
        assert!((median - 3.0).abs() < EPSILON);
    }

    #[test]
    fn test_median_even() {
        let data = vec![1.0, 2.0, 3.0, 4.0];
        let median = calculate_median(&data).unwrap();
        assert!((median - 2.5).abs() < EPSILON);
    }

    #[test]
    fn test_variance_population() {
        let data = vec![2.0, 4.0, 4.0, 4.0, 5.0, 5.0, 7.0, 9.0];
        let variance = calculate_variance(&data, false).unwrap();
        assert!((variance - 4.0).abs() < 0.1);
    }

    #[test]
    fn test_std_dev() {
        let data = vec![2.0, 4.0, 4.0, 4.0, 5.0, 5.0, 7.0, 9.0];
        let std_dev = calculate_std_dev(&data, false).unwrap();
        assert!((std_dev - 2.0).abs() < 0.1);
    }

    #[test]
    fn test_range() {
        let data = vec![1.0, 5.0, 3.0, 9.0, 2.0];
        let range = calculate_range(&data).unwrap();
        assert!((range - 8.0).abs() < EPSILON);
    }
}
